shadow subscriber fatals the server

The server for some reason added more than one shadow subscriber to its internal state. This makes it impossible for the server to start.

`
[12176] 2017/06/12 13:42:29.336373 [INF] STREAM: Failed to start: there should be only one shadow subscriber for ["backend-durable:remove-bubble-group"] queue group
panic: there should be only one shadow subscriber for ["backend-durable:remove-bubble-group"] queue group [recovered]
	panic: there should be only one shadow subscriber for ["backend-durable:remove-bubble-group"] queue group
`

With some printf, I found:
- [client id: ]: ID:5 qGroup:"backend-durable:remove-bubble-group" inbox:"_INBOX.M3V34IJY91XSXFTB3G0NIQ" ackInbox:"_INBOX.sYCnOp4GzXNhtBPSPmJr3J" maxInFlight:1 ackWaitInSecs:1800 lastSent:31 isDurable:true 
[12176] 2017/06/12 13:42:29.335988 [INF] STREAM: debug 

- [client id: ]: ID:6 qGroup:"backend-durable:remove-bubble-group" inbox:"_INBOX.F8HZ11KI1TANGB68XPHATU" ackInbox:"_INBOX.bB3pce3nTYX4KGZy9uDjMJ" maxInFlight:1 ackWaitInSecs:1800 lastSent:31 isDurable:true

--- a/server/server.go
+++ b/server/server.go
@@ -408,11 +408,15 @@ func (ss *subStore) updateState(sub *subState) {
 		// The recovered shadow queue sub will have ClientID=="",
 		// keep a reference to it until a member re-joins the group.
 		if sub.ClientID == "" {
-			// Should not happen, if it does, panic
-			if qs.shadow != nil {
-				panic(fmt.Errorf("there should be only one shadow subscriber for [%q] queue group", sub.QGroup))
+			// There should be only one shadow queue subscriber, but
+			// we found in https://github.com/nats-io/nats-streaming-server/issues/322
+			// that some datastore had 2 of those (not sure how this happened except
+			// maybe due to upgrades from much older releases that had bugs?).
+			// So don't panic and use as the shadow the one with the highest LastSent
+			// value.
+			if qs.shadow == nil || sub.LastSent > qs.lastSent {
+				qs.shadow = sub
 			}
-			qs.shadow = sub
 		} else {
 			qs.subs = append(qs.subs, sub)
 		}

--- a/server/server_test.go
+++ b/server/server_test.go
@@ -23,6 +23,7 @@ import (
 	"github.com/nats-io/go-nats"
 	"github.com/nats-io/go-nats-streaming"
 	"github.com/nats-io/go-nats-streaming/pb"
+	"github.com/nats-io/nats-streaming-server/spb"
 	"github.com/nats-io/nats-streaming-server/stores"
 	"github.com/nats-io/nuid"
 )
@@ -6293,5 +6294,64 @@ func TestSingleConfFileForBoth(t *testing.T) {
 		cmd.Process.Kill()
 		cmd.Wait()
 		cmd = nil
+	}
+}
+
+func TestFileStoreMultipleShadowQSubs(t *testing.T) {
+	cleanupDatastore(t, defaultDataStore)
+	defer cleanupDatastore(t, defaultDataStore)
+
+	opts := getTestDefaultOptsForFileStore()
+	s := runServerWithOpts(t, opts, nil)
+	s.Shutdown()
+
+	fs, err := stores.NewFileStore(defaultDataStore, &stores.DefaultStoreLimits)
+	if err != nil {
+		t.Fatalf("Error creating store: %v", err)
+	}
+	defer fs.Close()
+	cs, _, err := fs.CreateChannel("foo", nil)
+	if err != nil {
+		t.Fatalf("Error creating channel: %v", err)
+	}
+	sub := spb.SubState{
+		ID:            1,
+		AckInbox:      nats.NewInbox(),
+		Inbox:         nats.NewInbox(),
+		AckWaitInSecs: 30,
+		MaxInFlight:   10,
+		LastSent:      1,
+		IsDurable:     true,
+		QGroup:        "dur:queue",
+	}
+	cs.Subs.CreateSub(&sub)
+	sub.ID = 2
+	sub.LastSent = 2
+	cs.Subs.CreateSub(&sub)
+	fs.Close()
+
+	// Should not panic
+	s = runServerWithOpts(t, opts, nil)
+	defer s.Shutdown()
+	scs, err := s.lookupOrCreateChannel("foo")
+	if err != nil {
+		t.Fatalf("Error looking up channel: %v", err)
+	}
+	ss := scs.UserData.(*subStore)
+	ss.RLock()
+	qs := ss.qsubs["dur:queue"]
+	ss.RUnlock()
+	if qs == nil {
+		t.Fatal("Should have recovered queue group")
+	}
+	qs.RLock()
+	shadow := qs.shadow
+	lastSent := qs.lastSent
+	qs.RUnlock()
+	if shadow == nil {
+		t.Fatal("Should have recovered a shadow queue sub")
+	}
+	if shadow.ID != 2 || lastSent != 2 {
+		t.Fatalf("Recovered shadow queue sub should be ID 2, lastSent 2, got %v, %v", shadow.ID, lastSent)
 	}
 }

